src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs

```diff
@@ -3082,8 +3082,20 @@ private IEnumerable<PSModuleInfo> CreateFakeModuleObject(IEnumerable<ModuleSpeci
                 // In that case, the nested module will first be loaded with a different session state, and then when trying to load the RootModule via 'LoadModuleNamedInManifest',
                 // the same loaded nested module will be reused for the RootModule by 'LoadModuleNamedInManifest'.
 
-                // Change the module name to match the manifest name, not the original name
+                // Initially the original module name comes from RootModule manifest property but real name comes from the module (manifest) path.
+                // Change the module name to match the manifest name, not the original name.
+                // Then do the same for all providers loaded in the module since a provider caches full name based on the module name.
                 newManifestInfo.SetName(manifestInfo.Name);
+                if (ss != null)
+                {
+                    foreach (ProviderInfo pi in ss.Provider.GetAll())
+                    {
+                        if (object.ReferenceEquals(pi.Module, newManifestInfo))
+                        {
+                            pi.SetModule(manifestInfo);
```

Comment:
I think we probably should built in the start-ness into `ProviderInfo.FullName` instead to change the cached `_fullName` when the `ModuleName` property is changing.
Something like this:
```
private string _cachedModuleName = null;
internal string FullName
{
    get
    {
        string GetFullName(string name, string psSnapInName, string moduleName)
        { ... }

        if (_fullName != null && _cachedModuleName == ModuleName)
        {
            return fullName;
        }

        _cachedModuleName = ModuleName;
        return (_fullName = GetFullName(Name, PSSnapInName, ModuleName));
    }
}
```